請設計一個名為 add的add() 函式，計算兩個數的和，並回傳到主程式。

輸入範例:
10
45

輸出範例: 
55

In [3]:
'''
21_1.自訂函式-相加(5.0)
'''
def add(x,y):
      sum=x+y
      return sum

x=int(input())
y=int(input())
print(add(x,y))

55


根據python函式的定義，自訂一個名為hello的函式，印出Hello function。

輸入範例:

輸出範例: 
Hello function

In [1]:
'''
21_2.自訂函式-Hello function(5.0)
'''
def hello():
    print("Hello function")

hello()

Hello function


根據python函式的定義，自訂一個名為hello的函式，印出Hello function，此函式可以印出n次的Hello function。

輸入範例:
10
輸出範例: 
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function

In [4]:
'''
21_3.自訂函式-Hello funtion 10次(5.0)
'''
def hello(n):
    for i in range(1,1+n,1):
        print("Hello function")

n=int(input())
hello(n)

Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function
Hello function


你會算1到100的和嗎?有甚麼方法呢?

方法1：使用梯形面積公式，(上底+下底)*高/
方法2：使用迴圈從1開始每一次遞增1，然後逐筆累加。
除了以上兩種方法，我們要介紹函式的方法，來解決這個問題。
請設計一個名為 add的add() 函式，計算1加到100的和，並回傳到主程式。

輸入範例:

輸出範例: 
5050

In [5]:
'''
21_4.自訂函式-累加(5.0)
'''
def add():
      sum=0
      for i in range(1, 101, 1):
           sum=sum+i
      return sum
print(add())

5050


你會算1到n的和嗎?有甚麼方法呢?

方法1：使用梯形面積公式，(上底+下底)*高/
方法2：使用迴圈從1開始每一次遞增1，然後逐筆累加。
除了以上兩種方法，我們要介紹以函式的方法，來解決這個問題。
請設計一個名為 add的add() 函式，計算1加到n的和，並回傳到主程式，本題請先使用自訂函式add設計方法1的程式。

輸入範例:
10

輸出範例: 
55

In [6]:
'''
21_5.自訂函式-累加-梯形公式(5.0)
'''
def add(x):
  sum =int((1+x)*x/2)
  return sum

x=int(input())
print(add(x))

55


你會算1到n的和嗎?有甚麼方法呢?

方法1：使用梯形面積公式，(上底+下底)*高/
方法2：使用迴圈從1開始每一次遞增1，然後逐筆累加。
除了以上兩種方法，我們要介紹已函式的方法，來解決這個問題。
請設計一個名為 add的add() 函式，使用方法2設計自訂函式，並回傳到主程式。

輸入範例:
10

輸出範例: 
55

In [7]:
'''
21_6.自訂函式-相加-累加(10.0)
'''
def add(n):
    total = 0
    for i in range(1,n+1):
        total += i
    return total


# 主程式
if __name__ == "__main__":    
    n = int(input())
    print(add(n))

55


請用自訂一個名字含有area的函式解題。

計算高 10，半徑 3 的圓柱體體積(圓周率=3.14)是多少？
以及當半徑多加 2 後，圓柱體體積又會變成是多少？

請注意，如果沒有依照上述規則，將會產生無法判分的情形。

輸出範例: 
半徑3時的圓柱體體積: 282.6
半徑5時的原柱體體積: 785.0

In [8]:
'''
21_7.自訂函式-圓柱體體積(10.0)
'''
def area3(r, h):
      area2=r*r*3.14 # 圓面積公式
      return area2 * h # 圓柱體公式
r = 3
volume = area3(r, 10)
print('半徑{0}時的圓柱體體積: {1}'.format(r, volume))
r = 3+2
volume = area3(r, 10)
print('半徑{0}時的圓柱體體積: {1}'.format(r, volume))

半徑3時的圓柱體體積: 282.6
半徑5時的圓柱體體積: 785.0


設計一個名為volume的自訂函式，計算圓柱體的體積：輸入半徑與高，計算圓柱體體積(圓周率=3.14)。

使用自定函式的方法，程式會先呼叫一個圓柱體體積的函式 volume()， 再由 volume() 呼叫計算面積函式 area() 而完成體積計算，程式可以重複執行 10 次。

雖然有點煩，但是請用這個簡單的例子，體驗將大問題切成小問題的方法。

輸入範例: 
3
4
5
5
6
6
11
15
6
6
73
4
40
4
6
8
8
9
10
10

輸出範例: 
圓柱體體積113.04
圓柱體體積392.50
圓柱體體積678.24
圓柱體體積5699.10
圓柱體體積678.24
圓柱體體積66932.24
圓柱體體積20096.00
圓柱體體積904.32
圓柱體體積1808.64
圓柱體體積3140.00

In [9]:
'''
21_8.自訂函式-圓面積與體積(10.0)
'''
def area(radius):
    pi = 3.14
    return radius ** 2 * pi
    
def volume(radius,height): # 採用半徑與高度的函式
    return area(radius) * height
    
if __name__ == "__main__":
    for i in range(10):
        radius = int(input())
        height = int(input())
        print(f"圓柱體體積{volume(radius, height):.2f}")


圓柱體體積113.04
圓柱體體積392.50
圓柱體體積678.24
圓柱體體積5699.10
圓柱體體積678.24
圓柱體體積66932.24
圓柱體體積20096.00
圓柱體體積904.32
圓柱體體積1808.64
圓柱體體積3140.00


請設計一個名為madepow的函式解決以下問題：

計算 x 的 y 次方，直到 x=0 為止。

輸入範例: 
3
2
4
3
10
2
5
4
6
3
0

輸出範例: 
9
64
100
625
216

In [10]:
'''
21_9.自訂函式-x的y次方(10.0)
'''
def madepow():
    while True:
        x = int(input())
        if x == 0 :
            break
        y = int(input())
        print(x**y)
        
if __name__ == "__main__":
    madepow()


9
64
100
625
216


使用自訂函式的方法，自訂名為crazyprint的函式解題，印出為什麼需要副程式，方便我瘋狂列印。 輸入一個整數 N，決定列印次數。

輸入範例: 
10

輸出範例: 
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。

In [11]:
'''
21_10.自訂函式-無傳回值函數(10.0)
'''
def crazyprint():
    N = int(input())
    for i in range(N):
        print("為什麼需要副程式，方便我瘋狂列印。")

if __name__ == "__main__":
    crazyprint()

為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。
為什麼需要副程式，方便我瘋狂列印。


請自定名為prime的函式來解題。

質數的定義為：除了 1 和它本身之外，沒有別的數可以整除它的。

給一個整數代表測試資料筆數，判斷每一個數是否為質數。

輸入範例: 
10(代表次數，以10為例)
409
37
311
211
84
96
91
241
311
283

輸出範例:
質數
質數
質數
質數
不是質數
不是質數
不是質數
質數
質數
質數

In [12]:
'''
21_11.自訂函式-質數判斷(15.0)
'''
my_list=[] # 創建空的列表

def prime():
    # 測試資料筆數
    test = int(input())
    # 輸入測試資料
    for _ in range(test):
        n = int(input())
        my_list.append(n)
    
    # 判斷是否為質數
    for num in my_list:
        if num > 1 : # 質數必須大於1
            is_prime = True
            # 非質數必為乘積n = a*b 所以a最大值不會超過n**0.5
            for i in range(2,int(num ** 0.5) + 1) : 
                if num % i == 0:
                    is_prime = False # 如果不是質數 則跳出迴圈
                    break
            if is_prime:
                print("質數")
            else:
                print("不是質數")
        # 小於1不是質數
        else:
            print("不是質數")
                    
if __name__ == "__main__":
    prime()

質數
質數
質數
質數
不是質數
不是質數
不是質數
質數
質數
質數


請自訂 名為countprime的 函式練習(本題有限制要求，未依此規定不法判分)。

質數的定義為：除了 1 和它本身之外，沒有別的數可以整除它的。

給一個整數，計算小於等於此數的質數個數，程式可以重複執行直到小於 1 為止。
輸入到小於 2 時結束。

輸入範例:
2
3
4
5
6
7
8
9
10
30
40
50
60
1

輸出範例: 
1
2
2
3
3
4
4
4
4
10
12
15
17

In [13]:
'''
21_12.自訂函式-質數個數(10.0)
'''

def countprime(num):
    count = 0 # 計數器
    # 判斷是否為質數
    for num in range(2,num+1,1):
        is_prime = True
        # 非質數必為乘積n = a*b 所以a最大值不會超過n**0.5
        for i in range(2,int(num ** 0.5) + 1) : 
            if num % i == 0:
                is_prime = False
                break
        if is_prime:
            count += 1
    return count

if __name__ == "__main__":
    while True :
        try:
            num = int(input())
            if num < 2:
                break
            print(countprime(num))  # 輸出質數個數
        except ValueError :
            print("please input correct num again!")

1
2
2
3
3
4
4
4
4
10
12
15
17


請用自訂函式 countprime() 解題。

質數的定義為：除了 1 和它本身之外，沒有別的數可以整除它的。

你的程式必須輸入兩正整數 A、B，請傳回 A~B 之間(包含A、B)有多少個質數。 程式需重複執行，直到輸入 A 大於 B 為止。

輸入範例: 
1 10
10 20
20 30
30 40
2 1

輸出範例: 
4
4
2
2

In [14]:
'''
21_13.自訂函式-彈：也是質數個數(5.0)
'''
def countprime():
    while True:
        try:
            # map() 是一個把函式「套用」到每個元素上的工具。
            A, B = map(int, input().split())  # 同一行讀取 A 和 B  輸入時需空一格
            if A > B:
                break
            count = 0
            for i in range(A, B+1):
                if i < 2:
                    continue # < 2 的不是質數
                for j in range(2,int(i**0.5)+1):  # 判斷是否為質數（只檢查到平方根）
                    if i % j == 0 :
                        break
                else:
                    count += 1
            print(count)
        except: #try-excepr 可捕捉ValueError
            break
countprime()

4
4
2
2


請用自訂名為prime的函式解題(本題有限制要求，未依此規定不法判分)。

質數的定義為：除了 1 和它本身之外，沒有別的數可以整除它的。

你的程式必須當輸入一正整數 M，請找出小於 M 的最大質數，並回傳該值。 程式需重複執行，直到輸入 -1 為止。

輸入範例: 
6
9
20
-1

輸出範例: 
5
7
19

In [15]:
'''
21_14.自訂函式-彈: 最大質數(5.0)
'''

def prime():
    try:
        while True:
            M = int(input())
            if M == -1 :
                break
            max_prime = -1  # 預設沒找到質數
            
            # 因為要找出<M的最大質數 所以範圍是2~M-1
            for i in range(2,M):
                for j in range(2,int(i**0.5)+1):  # 判斷是否為質數（只檢查到平方根）
                    if i % j == 0 :
                        break
                else:
                    max_prime = i # 每找到一個質數就更新最大值
            
            print(max_prime)
    except:
        print("error")
    
prime()

5
7
19
